src/osd/PeeringState.cc

```diff
@@ -1217,120 +1209,129 @@ void PeeringState::reject_reservation()
  *  2) Prefer longer tail if it brings another info into contiguity
  *  3) Prefer current primary
  */
-map<pg_shard_t, pg_info_t>::const_iterator PeeringState::find_best_info(
+std::optional<std::reference_wrapper<const std::pair<const pg_shard_t, pg_info_t>>>
+PeeringState::find_best_info(
   const map<pg_shard_t, pg_info_t> &infos,
   bool restrict_to_up_acting,
   bool *history_les_bound) const
 {
   ceph_assert(history_les_bound);
-  /* See doc/dev/osd_internals/last_epoch_started.rst before attempting
-   * to make changes to this process.  Also, make sure to update it
-   * when you find bugs! */
-  eversion_t min_last_update_acceptable = eversion_t::max();
-  epoch_t max_last_epoch_started_found = 0;
-  for (map<pg_shard_t, pg_info_t>::const_iterator i = infos.begin();
-       i != infos.end();
-       ++i) {
-    if (!cct->_conf->osd_find_best_info_ignore_history_les &&
-	max_last_epoch_started_found < i->second.history.last_epoch_started) {
+  // See doc/dev/osd_internals/last_epoch_started.rst before attempting
+  // to make changes to this process.  Also, make sure to update it
+  // when you find bugs!
+  auto min_last_update_acceptable = eversion_t::max();
+  epoch_t max_les = 0;
+  const bool ignore_history_les =
+    cct->_conf->osd_find_best_info_ignore_history_les;
+  for ([[maybe_unused]] auto& [shard, info] : infos) {
+    if (!ignore_history_les && max_les < info.history.last_epoch_started) {
       *history_les_bound = true;
-      max_last_epoch_started_found = i->second.history.last_epoch_started;
+      max_les = info.history.last_epoch_started;
     }
-    if (!i->second.is_incomplete() &&
-	max_last_epoch_started_found < i->second.last_epoch_started) {
+    if (!info.is_incomplete() &&
+        max_les < info.last_epoch_started) {
       *history_les_bound = false;
-      max_last_epoch_started_found = i->second.last_epoch_started;
+      max_les = info.last_epoch_started;
     }
   }
-  for (map<pg_shard_t, pg_info_t>::const_iterator i = infos.begin();
-       i != infos.end();
-       ++i) {
-    if (max_last_epoch_started_found <= i->second.last_epoch_started) {
-      if (min_last_update_acceptable > i->second.last_update)
-	min_last_update_acceptable = i->second.last_update;
+  for ([[maybe_unused]] auto& [shard, info] : infos) {
```

Comment:
If we're trying to make this more readable, perhaps s/info/candidate_info?

Reply:
done.